### PR TITLE
autograd gradients computed on the server-side

### PR DESCRIPTION
**Pull request by momchil-flex**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,6 +29,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Support for differentiation with respect to monitor attributes that require interpolation, such as flux and intensity.
 - Support for automatic differentiation with respect to `.eps_inf` and `.poles` contained in dispersive mediums `td.PoleResidue` and `td.CustomPoleResidue`.
 - Support for `FieldProjectionAngleMonitor` for 2D simulations with `far_field_approx = True`.
+- Introduce RF material library. Users can now export `rf_material_library` from `tidy3d.plugins.microwave`.
```
Comment (tylerflex) [inline CHANGELOG.md]: removing this line

File: CHANGELOG.md [hunk 2]
```diff
@@ -29,6 +29,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- `autograd` gradient calculations are done server side by default. Behavior can be changed by passing `local_gradient = True` into `web.run()`.
```
Comment (tylerflex) [inline CHANGELOG.md]: adding "and `web.run_async()`"

Comment (tylerflex) [reply CHANGELOG.md]: also moving to the Unreleased header.

File: tests/test_components/test_autograd.py [hunk 3]
```diff
@@ -586,6 +655,53 @@ def objective(*args):
         print(f"{num_structures_test} structures took {t2:.2e} seconds")
 
 
+@pytest.mark.parametrize("structure_key, monitor_key", (("custom_med", "mode"),))
```
Comment (tylerflex) [inline tests/test_components/test_autograd.py]: will change this to iterate over all of the structures and monitors

File: tests/test_components/test_autograd.py [hunk 4]
```diff
@@ -116,39 +117,107 @@
...
+                cache[task_id_fwd][AUX_KEY_SIM_FIELDS_KEYS] = sim_fields_keys
+                # return original data only
+                return aux_data[AUX_KEY_SIM_DATA_ORIGINAL], task_id_fwd
+            else:
+                return run_emulated(simulation, task_name=task_name), task_id_fwd
+
+        def emulated_run_bwd(simulation, task_name, **run_kwargs) -> td.SimulationData:
+            """What gets called instead of ``web/api/autograd/autograd.py::_run_tidy3d_bwd``."""
+
+            task_id_fwd = task_name[:-8]
```
Comment (tylerflex) [inline tests/test_components/test_autograd.py:168]: not my proudest line of code. There's probably a more reliable way to grab this from the cache?

File: tests/test_components/test_autograd.py [hunk 5]
```diff
@@ -635,57 +768,6 @@ def objective(*args):
         ag.grad(objective)(params0)
 
 
-def test_web_failure_handling(log_capture, monkeypatch, use_emulated_run, use_emulated_run_async):
```
Comment (tylerflex) [inline tests/test_components/test_autograd.py:638]: why was this deleted?

File: tidy3d/plugins/invdes/design.py [hunk 6]
```diff
@@ -129,8 +129,9 @@ def to_simulation_data(self, params: anp.ndarray, **kwargs) -> td.SimulationData
         """Convert the ``InverseDesign`` to a ``td.Simulation`` and run it."""
         simulation = self.to_simulation(params=params)
         kwargs.setdefault("task_name", self.task_name)
-        sim_data = web.run(simulation, verbose=self.verbose, **kwargs)
-        return sim_data
+        return web.run(simulation, verbose=self.verbose, **kwargs)
+        # sim_data = job.run()
```
Comment (tylerflex) [inline tidy3d/plugins/invdes/design.py:133]: removing comments

File: tidy3d/web/api/connect_util.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
 from ...log import log
 
 # number of seconds to keep re-trying connection before erroring
-CONNECTION_RETRY_TIME = 180
+CONNECTION_RETRY_TIME = 0.180
```
Comment (tylerflex) [inline tidy3d/web/api/connect_util.py]: will fix this

File: tidy3d/web/api/autograd/utils.py [hunk 8]
```diff
@@ -83,3 +63,46 @@ def get_field_key(dim: str, fld_data: typing.Union[td.FieldData, td.Permittivity
...
+        title="Collection of tracers.",
+    )
+
+    @property
+    def to_autograd_field_map(self) -> AutogradFieldMap:
+        """Convert to ``AutogradFieldMap`` autograd dictionary."""
+        return dict_ag({tracer.path: tracer.data for tracer in self.tracers})
+
+    @classmethod
+    def from_autograd_field_map(cls, autograd_field_map) -> "FieldMap":
```
Comment (tylerflex) [inline tidy3d/web/api/autograd/utils.py]: will use `__future__.annotations` for this

File: tidy3d/web/api/autograd/autograd.py [hunk 9]
```diff
@@ -29,20 +38,8 @@
...
-
-    td.log.warning(
-        f"Autograd compatible '{fn_name}' failed, running original '{fn_name}'. "
-        "If you received this warning, please file an issue at the tidy3d front end with this "
-        f"message pasted in and we will investigate. \n\n "
-        f"{URL_LINK}.\n\n"
-        f"{exc_str} {traceback_str}.\n\n"
-    )
+# default value for whether to do local gradient calculation (True) or server side (False)
+LOCAL_GRADIENT = True
```
Comment (tylerflex) [inline tidy3d/web/api/autograd/autograd.py:40]: should we make this default of False?

**Review by tylerflex — COMMENTED** [review]
Added my comments for reference (which I fixed in my last two commits)

File: tests/test_components/test_autograd.py (hunk 4 above)
Comment (momchil-flex) [reply tests/test_components/test_autograd.py:168]: Yeah I winced at that too. It's related to [this](https://github.com/flexcompute/tidy3d/blob/3a8c20972eb6eb3f19495b912f9a0d51d1bd62cb/tidy3d/web/api/autograd/autograd.py#L685-L692) which is also not that pretty. No obvious way to improve but I think fundamentally an internal rethink may be needed rather than specifically fixing the cache grabbing in the test.

File: tests/test_components/test_autograd.py (hunk 5 above)
Comment (momchil-flex) [reply tests/test_components/test_autograd.py:638]: Because [this](https://github.com/flexcompute/tidy3d/pull/1824/files#diff-9f6ac729a84eb7e5b2ebe88ff4a4b52e63a0e21ec3e47d0553a9d0c1298b1a77L33-L45) was removed, but maybe the test needs to be modified instead?
